### PR TITLE
[test gap] Add long jump behavior test

### LINKED ISSUE #3591: Need new test: System and Management - Network Time Protocol (NTP)
This is a test gap:

Long jump behavior control

### PR DESCRIPTION
**Pull request by diaryevil**
Signed-off-by: Yuxuan Ye <yuxuanye@microsoft.com>

### Description of PR

Add long jump behavior test for test_ntp

Summary:
Fixes #3591

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911

#### What is the motivation for this PR?

Add long jump behavior test for test_ntp

#### How did you do it?

Set the DUT's clock forward for one hour, and wait for it to sync

#### How did you verify/test it?

Test it on vs and a DUT with image 202012

### CODE REVIEW

**Review by yxieca — DISMISSED** [review]
Test should first set the NTP configuration to allow or disallow NTP long jump. Then check the behavior to be consistent with the expected behavior.

We know that even though we disabled long jump, the long jump still going to happen. So one test case (allow long jump) should pass, and the other testcase (disallow long jump) should fail (xfail).

File: tests/ntp/test_ntp.py [hunk 1]
```diff
@@ -42,13 +55,58 @@ def setup_ntp(ptfhost, duthosts, rand_one_dut_hostname, creds):
...
+    duthost = duthosts[rand_one_dut_hostname]
+
+    duthost.service(name='ntp', state='stopped')
+    setup_long_jump_config(duthost, enable=True)
+    duthost.service(name='ntp', state='restarted')
+
+    pytest_assert(wait_until(720, 10, 0, check_ntp_status, duthost),
+                  "NTP long jump enable failed")
+
+@pytest.mark.xfail
```
[the remark this answers is not in the document]
Comment (diaryevil) [reply tests/ntp/test_ntp.py]: fixed

**Review by wangxin — APPROVED** [review]

**Comment by wangxin** [comment]
@yuxuanye Can you address the merge conflicts?

**Comment by diaryevil** [comment]
Conflict resloved

File: tests/common/plugins/conditional_mark/tests_mark_conditions.yaml [hunk 2]
```diff
@@ -464,3 +464,9 @@ platform_tests/api/test_thermal.py::TestThermalApi::test_set_high_threshold:
     reason: "Unsupported platform API"
     conditions:
       - "asic_type in ['mellanox']"
+
+ntp/test_ntp.py::test_ntp_long_jump_disabled:
+  # Due to NTP code bug, long jump will still happen after disable it.
+  # Set xfail flag for this test case
+  xfail:
```
Comment (yxieca) [inline tests/common/plugins/conditional_mark/tests_mark_conditions.yaml:471]: Please add struct: True

File: tests/ntp/test_ntp.py [hunk 3]
```diff
@@ -11,9 +12,21 @@
...
+def config_long_jump(duthost, enable=False):
+    """change ntpd option to enable or disable long jump"""
+    if enable:
+        logger.info("enable ntp long jump")
+        regex = "s/NTPD_OPTS='-x'/NTPD_OPTS='-g'/"
+    else:
+        logger.info("disable ntp long jump")
+        regex = "s/NTPD_OPTS='-g'/NTPD_OPTS='-x'/"
+
+    duthost.command("sed -i %s /etc/default/ntp" % regex)
```
Comment (yxieca) [inline tests/ntp/test_ntp.py:26]: You need to restart ntp service to complete the configuration.

File: tests/ntp/test_ntp.py [hunk 4]
```diff
@@ -42,13 +55,53 @@ def setup_ntp(ptfhost, duthosts, rand_one_dut_hostname, creds):
...
+    # stop NTP and set time on DUT
+    duthost.service(name='ntp', state='stopped')
+    duthost.command("date -s '@{}'".format(start_time_dut - TIME_FORWARD))
+
+    # set long jump config with variable
+    yield
+
+    # set DUT's time back after long jump test
+    duthost.service(name='ntp', state='stopped')
+    config_long_jump(duthost)
```
Comment (yxieca) [inline tests/ntp/test_ntp.py:77]: Collect NTP enabled/disabled state before test and retore it here?

File: tests/common/plugins/conditional_mark/tests_mark_conditions.yaml (hunk 2 above)
Comment (diaryevil) [reply tests/common/plugins/conditional_mark/tests_mark_conditions.yaml:476]: fixed

File: tests/ntp/test_ntp.py (hunk 4 above)
Comment (diaryevil) [reply tests/ntp/test_ntp.py]: That's better

File: tests/common/plugins/conditional_mark/tests_mark_conditions.yaml [hunk 5]
```diff
@@ -464,3 +464,10 @@ platform_tests/api/test_thermal.py::TestThermalApi::test_set_high_threshold:
     reason: "Unsupported platform API"
     conditions:
       - "asic_type in ['mellanox']"
+
+ntp/test_ntp.py::test_ntp_long_jump_disabled:
```
Comment (yxieca) [inline tests/common/plugins/conditional_mark/tests_mark_conditions.yaml:468]: Sorry, one last thing, please move this condition above according to alphabetic order of test cases.

Comment (diaryevil) [reply tests/common/plugins/conditional_mark/tests_mark_conditions.yaml]: fixed

File: tests/ntp/test_ntp.py (hunk 3 above)
Comment (diaryevil) [reply tests/ntp/test_ntp.py:26]: Fixed, and the restarts next it were also removed

**Review by yxieca — APPROVED** [review]
Please fix the conditional mar

**Pull request merged**